lib/puppet/parser/compiler.rb

```diff
@@ -790,9 +790,16 @@ def sanitize_node(node)
   # Set the node's parameters into the top-scope as variables.
   def set_node_parameters
     node.parameters.each do |param, value|
+      # We don't want to set @topscope['environment'] from the parameters,
+      # instead we want to get that from the node's environment itself in
+      # case a custom node terminus has done any mucking about with
+      # node.parameters.
+      next if param.to_s == 'environment'
       # Ensure node does not leak Symbol instances in general
       @topscope[param.to_s] = value.is_a?(Symbol) ? value.to_s : value
     end
+    @topscope['environment'] = node.environment.name.to_s
```

Comment:
Is this going to result in a behavior change for users? I feel like this might be safer to merge this change into master, but I also don't have a ton of context on how frequently folks are relying on parameters to set the environment

Reply:
If they're setting `node.parameters['environment']` to be something different from what `node.environment.name.to_s` is, then yes, it'll be a behavior change. Though, it's a behavior change in the sense that all bug fixes are behavior changes. `$::environment` is supposed to be the same as `node.environment.name.to_s` during catalog compilation as `node.environment` is the environment used for compilation (not `node.parameters['environment']`).

In the case where `node.environment` is `nil` and they're setting `node.parameters['environment']`, the "normal" lookup behavior of the `Node#environment` method will still use `node.parameters['environment']` to lookup the environment, so nothing will change there.
https://github.com/puppetlabs/puppet/blob/3b3f4a81d3f5fa05e31edb7bfc27e0891aa88fd0/lib/puppet/node.rb#L60-L78

Reply:
Should a test be added to ensure that the original behavior is preserved in the case where `node.environment` is nil and they're setting `node.parameters['environment']`?

Reply:
That is already handled by tests in `node_spec.rb`. Eg:
https://github.com/puppetlabs/puppet/blob/917849f65340753e6d50e13e4edb0d8fff23cf09/spec/unit/node_spec.rb#L30-L32